Allow including access lists in other access lists

lxea opened the pull request:
Allows recursive inclusion of Access Lists in other Access Lists.

For example, with the hierarchy below, `user A` would inherit traits and roles from `acl C`, `acl B`, and `acl A`, while `user B` would traits and roles from `acl C` and `acl B`:
```mermaid
flowchart TD
    A[acl A] --has member-->B(user A)
    C[acl B] --has member-->A
    C --has member--> D(user B)
    E[acl C] --has member-->C
```

changelog: Allow nested inclusion of Access Lists as Members and Owners in other Access Lists

RFD: https://github.com/gravitational/teleport/blob/master/rfd/0170-nested-accesslists.md

E: https://github.com/gravitational/teleport.e/pull/4549

Depends on #47828.

justinas commented on api/proto/teleport/accesslist/v1/accesslist.proto:
```diff
@@ -71,6 +71,23 @@ message AccessListSpec {
+
+// ParentAccessList contains information about access lists included
+// as parents in an access list
+message ParentAccessList {
```
From the rest of the code, looks like the AL specified by this object is considered a child of the `AccessListSpec` it is in, not the parent? Might want to name this `AccessListRef` or similar maybe, a la `UserRef`.

https://github.com/gravitational/teleport/blob/58ced9cd6e698fa8cb10cf44751644aa6db9d4ab/api/proto/teleport/legacy/types/types.proto#L3573-L3577

justinas commented on api/types/accesslist/accesslist.go:
```diff
@@ -155,6 +155,14 @@ type Spec struct {
+
+	// MemberAccessLists is a list of AccessList ids that user membership
+	// should be fetched from
+	MemberAccessLists []ParentAccessList `json:"member_access_lists" yaml:"access_lists"`
```
YAML and JSON key names are mismatched

fspmarshall commented:
Is there an RFD or other design document for this feature somewhere?  My gut reaction to the idea of access list recursion (or cross-dependency of any kind) is that it seems risky and hard to reason about.

In general, I'm of the opinion that composition of access-control primitives should have a well-defined acyclic hierarchy wherever possible.  One of the key design goals of a good access-control system should be to mitigate/prevent human error, both on the part of the programmer and the end user.  Keeping the compositional structure as simple as possible to trace, reason about, and model is an important part of that.

lxea commented:
> Is there an RFD or other design document for this feature somewhere? My gut reaction to the idea of access list recursion (or cross-dependency of any kind) is that it seems risky and hard to reason about.
> 
> In general, I'm of the opinion that composition of access-control primitives should have a well-defined acyclic hierarchy wherever possible. One of the key design goals of a good access-control system should be to mitigate/prevent human error, both on the part of the programmer and the end user. Keeping the compositional structure as simple as possible to trace, reason about, and model is an important part of that.

There isn't an RFD but I'm happy to write one up. Keeping the hierarchy acyclic seems like it would be ideal, however I'm not sure if the services being integrated with (Azure Entra  groups @justinas) support cycles in their group hierarchies

justinas commented:
> There isn't an RFD but I'm happy to write one up. Keeping the hierarchy acyclic seems like it would be ideal, however I'm not sure if the services being integrated with (Azure Entra groups @justinas) support cycles in their group hierarchies

Entra (sadly) seems to support cycles. Generally we'd like to represent the imported data as faithfully as possible. *Maybe* we could assume that cycles are uncommon in real-life scenarios and refuse to process such data, but that decision would not be up to me I think 🙂  /cc @jakule

lxea commented:
> > There isn't an RFD but I'm happy to write one up. Keeping the hierarchy acyclic seems like it would be ideal, however I'm not sure if the services being integrated with (Azure Entra groups @justinas) support cycles in their group hierarchies
> 
> Entra (sadly) seems to support cycles. Generally we'd like to represent the imported data as faithfully as possible. _Maybe_ we could assume that cycles are uncommon in real-life scenarios and refuse to process such data, but that decision would not be up to me I think 🙂 /cc @jakule

Maybe having it require --force flag in the terminal/a warning in the ui, if the data being imported or acl being modified introduces a cycle would be useful? or maybe that would just be overhead

r0mant reviewed (commented):
@lxea @justinas @fspmarshall @jakule We do need to support nested access lists (both EntraID and Github can have nested groups/teams) but I agree that cycles ("A" is member of "B", "B" is a member of "C", "C" is a member of "A") add too much complexity in the system so let's prevent them and see if anyone complains. I can't come up with a reasonable real-world scenario where such group membership could be useful.

Let's also not call them "recursive". Let's call them "nested".

codingllama commented on api/proto/teleport/accesslist/v1/accesslist.proto:
```diff
@@ -71,6 +71,23 @@ message AccessListSpec {
+
+  // member_access_lists is a list of access lists that user
+  // membership should be fetched from
+  repeated AccessListRef member_access_lists = 12;
```
Are cycles allowed?

Eg: A.member_access_lists = [B], B.member_access_lists = [C], C.member_access_lists = [A].

Please document.

codingllama commented on api/proto/teleport/accesslist/v1/accesslist.proto:
```diff
@@ -71,6 +71,23 @@ message AccessListSpec {
   AccessListGrants owner_grants = 11;
+
+  // member_access_lists is a list of access lists that user
+  // membership should be fetched from
```
So exclusively user memberships are pulled, and nothing else? I think it's worth being extra explicit in the language here.

Are the membership_requires of the parent AccessList applied before calculating the members?

Are the membership_requires of the present AccessList further applied to the members of the "members lists"?

Please document.

codingllama commented on api/proto/teleport/accesslist/v1/accesslist.proto:
```diff
@@ -71,6 +71,23 @@ message AccessListSpec {
+
+  // owner_access_lists is a list of access lists that owner
+  // membership should be fetched from
+  repeated AccessListRef owner_access_lists = 13;
```
Same questions as above. Please document here too.

codingllama reviewed (commented):
Drive-by review, as this is related to other stuff I was just looking at. Added a couple questions.

r0mant commented:
@lxea Is this implementation ready for re-review now that the RFD is approved, have you applied all changes discussed in the RFD?

lxea commented:
> @lxea Is this implementation ready for re-review now that the RFD is approved, have you applied all changes discussed in the RFD?

Yeah I think it should be good for re-review

lxea commented:
Think its good for re-review 👍

r0mant reviewed (commented):
@lxea Is this ready for another review?

r0mant commented on tool/tctl/common/acl_command.go:
```diff
@@ -70,6 +72,7 @@ func (c *ACLCommand) Initialize(app *kingpin.Application, _ *servicecfg.Config)
 	users := acl.Command("users", "Manage user membership to access lists.")
 
 	c.usersAdd = users.Command("add", "Add a user to an access list.")
+	c.usersAdd.Flag("refacl", "Add a reference to an access list to another access list").BoolVar(&c.isRefAcl)
```
Instead of a boolean flag, I would use the same `--type=user|list` switch as on the backend.

r0mant commented on tool/tctl/common/acl_command.go:
```diff
@@ -78,6 +81,7 @@ func (c *ACLCommand) Initialize(app *kingpin.Application, _ *servicecfg.Config)
 	c.usersRemove = users.Command("rm", "Remove a user from an access list.")
 	c.usersRemove.Arg("access-list-name", "The access list name.").Required().StringVar(&c.accessListName)
 	c.usersRemove.Arg("user", "The user to remove from the access list.").Required().StringVar(&c.userName)
+	c.usersRemove.Flag("refacl", "Add a reference to an access list to another access list").BoolVar(&c.isRefAcl)
```
Same comment as above.

r0mant commented on tool/tctl/common/acl_command.go:
```diff
@@ -234,20 +247,24 @@ func displayAccessLists(format string, accessLists ...*accesslist.AccessList) er
 
 func displayAccessListsText(accessLists ...*accesslist.AccessList) error {
-	table := asciitable.MakeTable([]string{"ID", "Review Frequency", "Review Day Of Month", "Granted Roles", "Granted Traits"})
+	table := asciitable.MakeTable([]string{"ID", "Review Frequency", "Review Day Of Month", "Granted Roles", "Granted Traits", "Nested Owner Lists"})
```
This feels a bit random: why do we show nested owner lists here when we don't display normal list owners or members?

r0mant reviewed (commented):
Just a few preliminary comments for now.

smallinsky commented on lib/services/access_list.go:
```diff
@@ -140,6 +140,10 @@ type AccessListMemberGetter interface {
 	// enough information to compute the dynamic member record.
 	GetAccessListMember(ctx context.Context, accessList string, memberName string) (*accesslist.AccessListMember, error)
+	// GetAccessList returns the specified access list resource.
+	GetAccessList(context.Context, string) (*accesslist.AccessList, error)
+	// GetAccessList returns the specified access list resource.
+	GetAccessLists(context.Context) ([]*accesslist.AccessList, error)
```
The interfaces name is `AccessListMemberGetter` 

By adding `GetAccessList` and `GetAccessLists` the interfaces name is obsolete.

smallinsky commented on lib/services/access_list.go:
```diff
@@ -174,6 +178,12 @@ type AccessListMembers interface {
 	DeleteAllAccessListMembers(ctx context.Context) error
 }
 
+type AccessListRecursiveMembersGetter interface {
+	ListAccessListMembers(ctx context.Context, accessListName string, pageSize int, pageToken string) (members []*accesslist.AccessListMember, nextToken string, err error)
+	GetAccessListMember(ctx context.Context, accessList string, memberName string) (*accesslist.AccessListMember, error)
+	GetAccessList(context.Context, string) (*accesslist.AccessList, error)
+}
```
How the `Recursive` logic applie to the interface methods ? 

I would assume that if I have `i AccessListRecursiveMembersGetter` and call i.GetAccessListMember the `GetAccessListMember` method on recursive interface will travel all nested access list and return the all members by visiting nested access list chain.

smallinsky commented on tool/tctl/common/acl_command.go:
```diff
@@ -234,20 +247,24 @@ func displayAccessLists(format string, accessLists ...*accesslist.AccessList) er
 			accessList.Spec.Audit.Recurrence.DayOfMonth.String(),
 			grantedRoles,
 			grantedTraits,
+			// todo(amk): include inherited granted roles/traits andmemberAccessLists,
```
Who is amk ?

smallinsky commented on lib/services/access_list.go:
```diff
@@ -243,20 +257,166 @@ func IsAccessListOwner(identity tlsca.Identity, accessList *accesslist.AccessLis
+					return trace.NotFound("error find access list %s", pal)
+				}
+				for _, next := range subAccessListMembers {
+					if _, ok := seen[next]; ok {
+						return trace.BadParameter("cycles are not allowed between dynamic access list members")
```
What about diamond hierarchy where there is no any cycles or other [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph)  ?: 
![Screenshot 2024-06-27 at 14 58 50](https://github.com/gravitational/teleport/assets/22402974/fb7e13e0-a612-4fe2-846e-2b44d69338dc)

smallinsky commented on lib/services/access_list.go:
```diff
@@ -243,20 +257,166 @@ func IsAccessListOwner(identity tlsca.Identity, accessList *accesslist.AccessLis
+			}
+		}
+	}
+	return trace.NotFound("user %s is not a member of the access list or its parents", username)
```
> "or its parents"

The logic travels  `access list  -> children` not `access list <- parents`

smallinsky reviewed (commented):
First iteration, I have left some comments. 


@lxea @r0mant 

Previously we had a security concerns that a access list  owner is able to escalate his privileges  by adding themselves as a member to the access list. 

To prevent that we have build a logic that prevent a owner to add themselves  [userTryingToAddThemselves](https://github.com/gravitational/teleport.e/blob/master/lib/accesslist/service.go#L915)

With dynamic ownership & membership we should prevent private escalation 


We should detect if an owner of an access list attempts to add themselves via a dynamic access list relation during member membership updates.

lxea commented on api/proto/teleport/accesslist/v1/accesslist.proto:
```diff
@@ -197,6 +214,12 @@ message MemberSpec {
   IneligibleStatus ineligible_status = 7;
+
+  // origin of the member, either "member", or "dynamic"
+  string origin = 9;
+
+  // access_list_title is the title of the access list being nested, used if origin is "dynamic"
+  string access_list_title = 10;
```
this is removed, also there is a thing at the top of the MemberSpec message that reserves 8, should we still use it?

kiosion replied:
@smallinsky friendly ping – is this change still needed?

lxea commented on lib/services/access_list.go:
```diff
@@ -243,20 +257,166 @@ func IsAccessListOwner(identity tlsca.Identity, accessList *accesslist.AccessLis
+				}
+				for _, next := range subAccessListMembers {
+					if _, ok := seen[next]; ok {
+						return trace.BadParameter("cycles are not allowed between dynamic access list members")
```
This is in a separate PR that is still in progress in e/ the upsert methods are all in there

r0mant commented on lib/services/access_list.go:
```diff
@@ -243,20 +257,166 @@ func IsAccessListOwner(identity tlsca.Identity, accessList *accesslist.AccessLis
+			return nil
+		},
+		func(current string) error {
+			member, err := members.GetAccessListMember(ctx, current, identity.Username)
```
Yes, when a list is nested (in either members or owners) we want to expand members of the list.

lxea commented on lib/services/access_list.go:
```diff
@@ -219,7 +229,7 @@ func UnmarshalAccessListMember(data []byte, opts ...MarshalOption) (*accesslist.
 
 // IsAccessListOwner will return true if the user is an owner for the current list.
-func IsAccessListOwner(identity tlsca.Identity, accessList *accesslist.AccessList) error {
+func IsAccessListOwner(ctx context.Context, members AccessListMemberGetter, identity tlsca.Identity, accessList *accesslist.AccessList) error {
```
I'll add documentation to make it clearer https://github.com/gravitational/teleport/pull/38738#discussion_r1662767998

smallinsky commented on lib/auth/userloginstate/generator_test.go:
```diff
@@ -554,6 +619,25 @@ func newAccessListMembers(t *testing.T, clock clockwork.Clock, accessList string
+func newAccessListMemberWithKind(t *testing.T, clock clockwork.Clock, accessList string, kind string, member string) *accesslist.AccessListMember {
+
+	var err error
+	res, err := accesslist.NewAccessListMember(header.Metadata{
```
the name is used to store access key member in backend

This is rare case but it can happen that a member `alice` exist where the alice is also name of access list.

That will result that the alice membership will be overwritten by the access list membership entry. 
 
What do you thing about decoupling  the name of the access list from the reference to the name of dynamic membership ?

lxea replied:
I think this makes sense. Am I right in saying this would only come up for manually created access list resources as I think all of them get uuids?

smallinsky commented on lib/services/access_list.go:
```diff
@@ -243,20 +258,147 @@ func IsAccessListOwner(identity tlsca.Identity, accessList *accesslist.AccessLis
+	seen := map[string]struct{}{}
+	queue := []string{accessList.GetName()}
+
+	for len(queue) > 0 {
+		size := len(queue)
+		for i := 0; i < size; i++ {
+			pal := queue[0]
+			queue = queue[1:]
```
Can we simplify this to: ? 
```suggestion
	for len(queue) > 0 {
		pal := queue[0]
		queue = queue[1:]
```

smallinsky commented on lib/services/access_list.go:
```diff
@@ -276,9 +418,21 @@ func (a AccessListMembershipChecker) IsAccessListMember(ctx context.Context, ide
+			return trace.NotFound("user %s is not a member of the access list", username)
+		}
+		if err != nil {
+			// Some other error has occurred
```
nit: The comment doesn't carry to much value for the reader can we remove it or explain the error flow ?

smallinsky commented on lib/services/access_list.go:
```diff
@@ -276,9 +418,21 @@ func (a AccessListMembershipChecker) IsAccessListMember(ctx context.Context, ide
+			// Some other error has occurred
+			return trace.Wrap(err)
+		}
+		return nil
```
The `UserMeetsRequirements` check in https://github.com/gravitational/teleport/pull/38738/files#diff-57e71bdbf9e192452bb9a2a45cb751930330be9ffe63caf97264c1c5aa324087R446 

is omitted so if the parent access list has some constraint about membership this check will be omitted. 
Resulting that access list constraint can be bypassed by nested access list behavior.

smallinsky reviewed (commented):
I have left few remarks but mostly looks reasonable to me.

lxea commented on lib/services/access_list.go:
```diff
@@ -243,20 +258,147 @@ func IsAccessListOwner(identity tlsca.Identity, accessList *accesslist.AccessLis
+
+}
+
+func recursiveIsAccessListOwnerCheck(ctx context.Context, members AccessListsAndMembersGetter, identity tlsca.Identity, accessList *accesslist.AccessList) error {
```
pushed a case however, isnt working correctly so I'll work a fix for ownership

kiosion commented:
> We should detect if an owner of an access list attempts to add themselves via a dynamic access list relation during member membership updates.

cc @smallinsky, looks like this was covered in the PR for enterprise [here](https://github.com/gravitational/teleport.e/blob/dab19c5f87a22541d6e9f77a7052136ed79fdca3/lib/accesslist/service.go#L914), did you have a chance to review that?

strideynet commented on lib/services/access_list.go:
```diff
@@ -243,20 +257,141 @@ func IsAccessListOwner(identity tlsca.Identity, accessList *accesslist.AccessLis
+
+		expires := member.Spec.Expires
+		if !expires.IsZero() && !a.clock.Now().Before(expires) {
+			return trace.AccessDenied("user %s's membership has expired in the access list", identity.Username)
```
I think there's a bug here, although, I think we'd want to check with someone what the expected behaviour should be:
- UserA is a member of AccessListA and AccessListB
- AccessListA and AccessListB are members of AccessListC
- UserA's membership of AccessListA is expired, but their membership of AccessListB is not.

It looks to me that when calculating membership of AccessListC:
- If AccessListA is processed first, then they receive this error.
- If AccessListB is processed first, then they are granted access

It probably ought to be deterministic and that will mean that even if we encounter an expired membership, we need to continue searching the tree.

kiosion commented on lib/services/access_list.go:
```diff
@@ -243,20 +257,141 @@ func IsAccessListOwner(identity tlsca.Identity, accessList *accesslist.AccessLis
+	return dynamicMembers, nil
+}
+
+func (a AccessListMembershipChecker) recursiveIsAccessListMemberCheck(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) error {
```
No, I also didn't see anything currently ensuring the max depth...

kiosion commented:
Rebased, fixed build errors, added some validation for nesting exceeding 10 levels, and got the e/frontend side of stuff working. Still need to add tests for all this & debug the role/trait inheritance – it seems to mostly work now, but there are some odd perm issues I ran into while doing some testing.

kiosion commented:
Should be good for a re-review

r0mant commented on lib/auth/authclient/api.go:
```diff
@@ -1144,7 +1144,7 @@ type Cache interface {
 
 	// CountAccessListMembers will count all access list members.
-	CountAccessListMembers(ctx context.Context, accessListName string) (uint32, error)
+	CountAccessListMembers(ctx context.Context, accessListName string) (uint32, uint32, error)
```
```suggestion
	CountAccessListMembers(ctx context.Context, accessListName string) (users uint32, lists uint32, err error)
```

r0mant commented on lib/auth/userloginstate/generator.go:
```diff
@@ -162,19 +163,23 @@ func (g *Generator) Generate(ctx context.Context, user types.User) (*userloginst
-// addAccessListsToState will added the user's applicable access lists to the user login state.
-func (g *Generator) addAccessListsToState(ctx context.Context, user types.User, state *userloginstate.UserLoginState) error {
+// addAccessListsToState will add the user's applicable access lists to the user login state.
+func (g *Generator) addAccessListsToState(ctx context.Context, user types.User, state *userloginstate.UserLoginState) (accesslist.Grants, error) {
```
This function is now a bit weird. It modifies login state and yet returns inherited grants for some reason. I understand you probably need them for the usage even emitted later. Can we add inherited grants to user state as well to keep logic consistent? Then you wouldn't need to modify any method signatures here either.

kiosion replied:
Yeah, I modified its return for use in the event. Could do to just add them to the state.

r0mant commented on lib/services/access_list.go:
```diff
@@ -218,47 +233,206 @@ func UnmarshalAccessListMember(data []byte, opts ...MarshalOption) (*accesslist.
-// IsAccessListMember will return true if the user is a member for the current list.
-func (a AccessListMembershipChecker) IsAccessListMember(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) error {
+// getAccessListDynamicMembers returns a list of all dynamic members (nested lists) for a given Access List.
+func getAccessListDynamicMembers(ctx context.Context, clt AccessListsAndMembersGetter, entry string) ([]string, error) {
```
Can we just say `getNestedAccessLists` then? Calling this "dynamic" is confusing, plus the godoc above actually has to explain what "dynamic" means.

r0mant commented on lib/services/access_list.go:
```diff
@@ -218,47 +233,206 @@ func UnmarshalAccessListMember(data []byte, opts ...MarshalOption) (*accesslist.
+	return dynamicMembers, nil
+}
+
+func (a AccessListMembershipChecker) recursiveIsAccessListMemberCheck(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) error {
```
Godoc here would be helpful. Similar to the owners check function below, to explain that the passed in identity is checked for membership.

r0mant commented on lib/services/access_list.go:
```diff
@@ -218,47 +233,206 @@ func UnmarshalAccessListMember(data []byte, opts ...MarshalOption) (*accesslist.
+	queue := []queueItem{{name: accessList.GetName(), depth: 0}}
+	membershipErr := trace.NotFound("user %s is not a member of the access list or its parents", identity.Username)
+
+	for len(queue) > 0 {
```
TBH I find the logic both here and in `recursiveIsAccessListOwnerCheck` (which is basically the same but for owners) slightly confusing. Similar to my comment below in `lib/services/local/access_list.go` - all this logic basically comes down to expanding either full owners or full members list, and then checking if the current identity is in there (plus the requirements check).

So with that in mind, can we have two helper methods: `ExpandAccessListOwners` and `ExpandAccessListMembers` that return full set of owners/members respectively? I provided an example of what an expand method would look like in that comment. Then you can use those helpers here and in the local service as well, and this expansion logic will be well-encapsulated and would lead to a more straightforward logic and probably better code reuse as well.

WDYT, can you refactor it like this?

kiosion replied:
I could give that a shot. By 'full set of members/owners', do you mean ignoring any membership/ownership requirements, and checking those separately?

kiosion replied:
If so, a refactor to `ExpandAccessListMembers` could work, but Owners don't have a reference to the Access List they're an owner of, so it might be more complicated to split out fetching them from verifying their ownership.

kiosion commented on lib/services/local/access_list.go:
```diff
@@ -138,6 +138,11 @@ func (a *AccessListService) GetAccessLists(ctx context.Context) ([]*accesslist.A
 
+// GetInheritedGrants returns grants inherited from parent access lists. This is not implemented in the local service.
+func (a *AccessListService) GetInheritedGrants(ctx context.Context, accessListID string) (*accesslist.Grants, error) {
+	return nil, trace.NotImplemented("GetInheritedGrants should not be called")
```
I think to satisfy the `AccessListService` interface, and since it's not used in the backend service? It's implemented in `api/client/accesslist/accesslist.go`:
https://github.com/gravitational/teleport/blob/3565893b08be6b0ec433b53cc50316ee86b971ba/api/client/accesslist/accesslist.go#L118-L130

kiosion commented on lib/services/local/access_list.go:
```diff
@@ -303,6 +319,73 @@ func (a *AccessListService) GetAccessListMember(ctx context.Context, accessList
 }
 
+// GetAccessListNestedOwners returns a list of all owners in an Access List with nested Access Lists as owners.
+func (a *AccessListService) GetAccessListNestedOwners(ctx context.Context, accessListName string) ([]*accesslist.Owner, error) {
```
Makes sense to me... would `GetFlattenedAccessListOwners` be clearer since `GetOwners()` exists on Access Lists?

r0mant replied:
I think just `GetAccessListOwners` would be fine. From the caller's perspective, it just wants to fetch all owners, I don't see an issue with a similarly named method on the list object.

kiosion commented on lib/services/local/access_list.go:
```diff
@@ -303,6 +319,73 @@ func (a *AccessListService) GetAccessListMember(ctx context.Context, accessList
+		// Owners are inherited from members of nested lists.
+		// So, on l0, we want to grab all the owners from the provided accessListName,
+		// while > l0, we want to fetch members of the nested lists.
+		if item.name == accessListName {
+			for _, owner := range accessList.GetOwners() {
+				if owner.MembershipKind == accesslist.MembershipKindList {
+					if _, ok := seen[owner.Name]; ok {
+						continue
+					}
+					seen[owner.Name] = struct{}{}
+					queue = append(queue, queueItem{name: owner.Name, depth: item.depth + 1})
+				} else {
+					allOwners = append(allOwners, &owner)
+				}
+			}
+		} else {
+			members, _, err := a.ListAccessListMembers(ctx, item.name, 0, "")
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			for _, member := range members {
+				if member.Spec.MembershipKind == accesslist.MembershipKindList {
+					if _, ok := seen[member.GetName()]; ok {
+						continue
+					}
+					seen[member.GetName()] = struct{}{}
+					queue = append(queue, queueItem{name: member.GetName(), depth: item.depth + 1})
+				} else {
+					allOwners = append(allOwners, &accesslist.Owner{
+						Name:             member.GetName(),
+						Description:      member.Metadata.Description,
+						MembershipKind:   member.Spec.MembershipKind,
+						IneligibleStatus: member.Spec.IneligibleStatus,
+					})
+				}
+			}
+		}
+	}
```
Replied to your other similar comment, but can give a shot. Would be nice to simplify some of the recursion in ownership/membership checks.

kiosion commented on web/packages/teleport/src/services/api/api.ts:
```diff
@@ -28,7 +28,7 @@ import parseError, { ApiError } from './parseError';
 
 const api = {
-  get(url, abortSignal?) {
+  get(url: string, abortSignal?: AbortSignal) {
```
Related to a change in the E PR; not sure why this wasn't a syntax error prior

kiosion commented on api/types/accesslist/accesslist.go:
```diff
@@ -299,6 +315,9 @@ func (a *AccessList) CheckAndSetDefaults() error {
 		if owner.Name == "" {
 			return trace.BadParameter("owner name is missing")
 		}
+		if owner.MembershipKind == "" {
```
That validation's handled in `e/lib/accesslist/service.go`, within `UpsertAccessListWithMembers` and `UpsertAccessListMember`

smallinsky commented on lib/auth/userloginstate/generator.go:
```diff
@@ -186,19 +191,31 @@ func (g *Generator) addAccessListsToState(ctx context.Context, user types.User,
 	for _, accessList := range accessLists {
-		if err := services.IsAccessListOwner(identity, accessList); err == nil {
-			g.grantRolesAndTraits(identity, accessList.Spec.OwnerGrants, state)
+		if dynamic, err := services.IsAccessListOwner(ctx, g.accessLists, identity, accessList); err == nil {
```
What is dynamic ? Can we rename it to isOwner ? 

Also you can simplify the logic by checking: 
```go
if isOwner, err := services.IsAccessListOwner(ctx, g.accessLists, identity, accessList); err == nil &&  {
   added := g.grantRolesAndTraits(identity, accessList.Spec.OwnerGrants, state)
   inheritedGrants.Roles = append(inheritedGrants.Roles, added.Roles...)
  for k, values := range added.Traits {
	  inheritedGrants.Traits[k] = append(inheritedGrants.Traits[k], values...)
  }
}
````

smallinsky commented on lib/services/access_list.go:
```diff
@@ -218,47 +233,206 @@ func UnmarshalAccessListMember(data []byte, opts ...MarshalOption) (*accesslist.
 	// An opaque access denied error.
 	accessDenied := trace.AccessDenied("access denied")
+	// Is the Ownership inherited, or explicit?
+	inherited := false
```
This is confusing. 

Function says that we are chakchecking if a user is IsAccessListOwner no matter if it is a top lever owner or owner by nested access list.

smallinsky commented on lib/auth/userloginstate/generator.go:
```diff
@@ -186,19 +191,31 @@ func (g *Generator) addAccessListsToState(ctx context.Context, user types.User,
-			g.grantRolesAndTraits(identity, accessList.Spec.OwnerGrants, state)
+		if dynamic, err := services.IsAccessListOwner(ctx, g.accessLists, identity, accessList); err == nil {
+			added := g.grantRolesAndTraits(identity, accessList.Spec.OwnerGrants, state)
+			if dynamic {
```
if dynamic is true it doesn't mean that a user is a owner/member vie nested access list The = `services.IsAccessListOwner` will return true for top parent lever owner/membership user  assignments.

smallinsky commented on lib/services/access_list.go:
```diff
@@ -218,47 +233,206 @@ func UnmarshalAccessListMember(data []byte, opts ...MarshalOption) (*accesslist.
+// IsAccessListOwner will determine if the user is an owner for the current list.
+// If ownership is inherited (through the user having Membership of a list which is an Owner of the provided `accessList`),
+// the user must also meet the ownership requirements of the parent list.
+func IsAccessListOwner(ctx context.Context, members AccessListsAndMembersGetter, identity tlsca.Identity, accessList *accesslist.AccessList) (bool, error) {
```
This function seems to be broken. 

if a user is top lever owner 

the following isOwner var will be set to true: 
```
isOwner := slices.ContainsFunc(accessList.GetOwners(), func(owner accesslist.Owner)
```

then we have following logic:
```
inherited := false
if !isOwner {
  inherited = true
}
return inherited
```

In the result in isOwner is true the inherited == false arg is returned from `IsAccessListOwner`

smallinsky reviewed (changes requested):
Let a few comments additional comments. 

The main issue is that the IsAccessListOwner seems to be broken, see: 
https://github.com/gravitational/teleport/pull/38738/files#diff-57e71bdbf9e192452bb9a2a45cb751930330be9ffe63caf97264c1c5aa324087R239

kiosion commented on integrations/access/accesslist/app.go:
```diff
@@ -249,10 +249,21 @@ func (a *App) getRecipientsRequiringReminders(ctx context.Context, accessList *a
+	var allOwners []*accesslist.Owner
+
+	allOwners, err := a.apiClient.GetAccessListNestedOwners(ctx, accessList.GetName())
+	if err != nil {
+		// TODO(kiosion): Remove in v18; protecting against server not having `GetAccessListNestedOwners` func.
```
Looking at this again, that call could theoretically error even if implemented - would it be best to just remove that comment and keep the fallback behaviour of logging the error + using the non-nested owners?

kiosion commented on integrations/access/accesslist/app.go:
```diff
@@ -249,10 +249,21 @@ func (a *App) getRecipientsRequiringReminders(ctx context.Context, accessList *a
-	allRecipients := make(map[string]common.Recipient, len(accessList.Spec.Owners))
+	var allOwners []*accesslist.Owner
+
+	allOwners, err := a.apiClient.GetAccessListNestedOwners(ctx, accessList.GetName())
```
Renamed it to `GetAccessListOwners` and clarified in the godoc for it that it returns all owners, including ones inherited from nested lists.

smallinsky commented on api/client/accesslist/accesslist.go:
```diff
@@ -113,6 +114,27 @@ func (c *Client) GetAccessListsToReview(ctx context.Context) ([]*accesslist.Acce
 }
 
+// GetInheritedGrants returns grants inherited by access list accessListID from parent access lists.
+func (c *Client) GetInheritedGrants(ctx context.Context, accessListID string) (accesslist.CombinedGrants, error) {
```
```suggestion
func (c *Client) GetInheritedGrants(ctx context.Context, accessListID string) (*accesslist.CombinedGrants, error) {
```

smallinsky commented on api/client/accesslist/accesslist.go:
```diff
@@ -113,6 +114,27 @@ func (c *Client) GetAccessListsToReview(ctx context.Context) ([]*accesslist.Acce
+		AccessListId: accessListID,
+	})
+	if err != nil {
+		return accesslist.CombinedGrants{}, trace.Wrap(err)
```
```suggestion
		return nil, trace.Wrap(err)
```

smallinsky commented on api/types/userloginstate/user_login_state.go:
```diff
@@ -53,6 +53,12 @@ type Spec struct {
 	// Traits are the traits attached to the user login state.
 	Traits trait.Traits `json:"traits" yaml:"traits"`
 
+	// InheritedRoles are any Roles that were inherited via inclusion in nested Access Lists.
+	InheritedRoles []string `json:"inherited_roles" yaml:"inherited_roles"`
+
+	// InheritedTraits are any Traits that were inherited via inclusion in nested Access Lists.
+	InheritedTraits trait.Traits `json:"inherited_traits" yaml:"inherited_traits"`
```
This object is stored in the backend side and right now it looks like these two properties were added only to emit usage audit event. 

I think that using a object that is stored on backend side just to forward few arguments from nested function is bad practice. 

The backend object should represent the Domain but right now the `InheritedRoles`, InheritedTraits where  we are adding a some garbage metadata to the UserLogicState. 

I would suggest collect and forward InheritedRoles and InheritedTraits directly in the code without using backend storage object.

smallinsky commented on lib/auth/userloginstate/generator.go:
```diff
@@ -186,19 +191,19 @@ func (g *Generator) addAccessListsToState(ctx context.Context, user types.User,
 }
 
-func (g *Generator) grantRolesAndTraits(identity tlsca.Identity, grants accesslist.Grants, state *userloginstate.UserLoginState) {
+func (g *Generator) grantRolesAndTraits(identity tlsca.Identity, grants accesslist.Grants, state *userloginstate.UserLoginState, membershipOrOwnershipType services.MembershipOrOwnershipType) accesslist.Grants {
```
Grants returned argument seems to be unused: 
```suggestion
func (g *Generator) grantRolesAndTraits(identity tlsca.Identity, grants accesslist.Grants, state *userloginstate.UserLoginState, membershipOrOwnershipType services.MembershipOrOwnershipType {
```

smallinsky commented on lib/services/local/access_list.go:
```diff
@@ -303,6 +320,84 @@ func (a *AccessListService) GetAccessListMember(ctx context.Context, accessList
+			var members []*accesslist.AccessListMember
+
+			for {
+				fetchedMembers, pageToken, err := a.ListAccessListMembers(ctx, item.name, 0, pageToken)
```
It looks like the page token logic is broken here 

note the on each following call:
```
fetchedMembers, pageToken, err := a.ListAccessListMembers(ctx, item.name, 0, pageToken)
```
the pageToken is scoped to for loop and reacted on each `ListAccessListMembers` call where the top 
```
var pageToken string
```

is never updated so `pageToken` injected to   `a.ListAccessListMembers(ctx, item.name, 0, pageToken)` always contains  `""`

in the result this logic creates a spin loop that existing backend resources.

r0mant replied:
@kiosion This is critical to fix, this class of issues with pagination has caused production outages and aborted deployments in Cloud before.

Also, can this be factored out into a helper method that returns all members with proper pagination? This function is long as-is.

kiosion replied:
The scoping issue w/ `pageToken` should be fixed.

Which parts would you want split into a helper? The following, which just fetches the paginated list of all members, or the whole second half of the func?

https://github.com/gravitational/teleport/blob/da585af4e9a2970ceedf9b5d8c9acd000e50f0c8/lib/services/local/access_list.go#L367-L380

r0mant replied:
Yeah I meant the helper for fetching all list members. We already have some example, even for access lists actually, maybe we can just export and reuse it? https://github.com/gravitational/teleport.e/blob/adb418b4ff297f036d3024116f38b1eb7f818121/lib/accesslist/reconciler_ineligible_status.go#L205-L220

smallinsky commented on lib/services/local/access_list.go:
```diff
@@ -303,6 +320,84 @@ func (a *AccessListService) GetAccessListMember(ctx context.Context, accessList
+				} else {
+					allOwners = append(allOwners, &owner)
+				}
+			}
+		} else {
```
can we simplify the loop and deflaten the `} else {
```suggestion
              continue
		}
```

so instead of 
```go
for _, v := elemst {
   if ... {
   } else {
     logic
   }
}

return all
```

we will have: 

```go
for _, v := elemst {
   if ... {
     continue
   }
   // logic
}

return all
```

smallinsky reviewed (commented):
Left few additional comments.

smallinsky commented on lib/services/local/access_list.go:
```diff
@@ -303,6 +320,84 @@ func (a *AccessListService) GetAccessListMember(ctx context.Context, accessList
+// GetAccessListOwners returns a list of all owners in an Access List, including those inherited from nested Access Lists.
+//
+// Returned Owners are not validated for ownership requirements – use `IsAccessListOwner` for validation.
+func (a *AccessListService) GetAccessListOwners(ctx context.Context, accessListName string) ([]*accesslist.Owner, error) {
```
Does this have any UT coverage ?

kiosion replied:
It should be indirectly tested via some other unit tests but I'll verify.

smallinsky reviewed (commented):
Left few additional comments.

r0mant commented on api/types/accesslist/convert/v1/accesslist.go:
```diff
@@ -209,11 +209,49 @@ func ToProto(accessList *accesslist.AccessList) *accesslistv1.AccessList {
 	}
 }
 
+func ToOwnerProto(owner accesslist.Owner) *accesslistv1.AccessListOwner {
```
Nit: Godocs please on this and `FromOwnerProto` below.

r0mant commented on integrations/access/accesslist/app.go:
```diff
@@ -249,10 +249,21 @@ func (a *App) getRecipientsRequiringReminders(ctx context.Context, accessList *a
+	var allOwners []*accesslist.Owner
+
+	allOwners, err := a.apiClient.GetAccessListOwners(ctx, accessList.GetName())
+	if err != nil {
+		// TODO(kiosion): Remove in v18; protecting against server not having `GetAccessListOwners` func.
+		log.WithError(err).Warnf("Error getting nested owners for access list %v, continuing with only explicit owners", accessList.GetName())
```
Should this check only for NotImplemented error?

kiosion replied:
Could do. Or do you think it'd be better to log errors here and try to continue with explicit owners only, vs. erroring on getting nested owners?

r0mant replied:
I would treat it as a transaction - you either get all owners or none of them if there was an error. Otherwise it's harder to reason about the logic and behavior could be confusing.

r0mant reviewed (commented):
@kiosion No major comments from my side apart from the feedback Marek left during the last review. The most critical thing to fix is the pagination issue pointed out [here](https://github.com/gravitational/teleport/pull/38738/files#r1774891174).

This PR has grown very big so let's avoid any further major refactorings or new additions (like, events, metrics, etc.), merge this so we can ship it and for other things do follow up PRs.

r0mant commented on lib/accesslists/hierarchy.go:
```diff
@@ -0,0 +1,569 @@
+	OwnerParents map[string]*HierarchyNode
+}
+
+type Hierarchy struct {
```
I would define `Hierarchy` interface so it's clear what it implements and rename this struct to `hierarchy`.

r0mant commented on lib/services/access_list.go:
```diff
@@ -218,47 +233,165 @@ func UnmarshalAccessListMember(data []byte, opts ...MarshalOption) (*accesslist.
+	var ownershipType MembershipOrOwnershipType
+
+	// Is the supplied identity explicitly in the Owners list?
+	if slices.ContainsFunc(accessList.GetOwners(), func(owner accesslist.Owner) bool {
+		return owner.Name == identity.Username
+	}) {
+		ownershipType = MembershipOrOwnershipTypeExplicit
+	} else {
+		err := a.recursiveIsAccessListOwnerCheck(ctx, identity, accessList)
+		if err != nil {
+			if trace.IsNotFound(err) {
+				return MembershipOrOwnershipTypeNone, trace.AccessDenied("User '%s' is not an owner of list '%s'.", identity.Username, accessList.GetName())
+			}
+			return MembershipOrOwnershipTypeNone, trace.Wrap(err)
+		}
+		ownershipType = MembershipOrOwnershipTypeInherited
+	}
+
+	// Does the supplied Identity meet the ownership requirements?
+	if !UserMeetsRequirements(identity, accessList.Spec.OwnershipRequires) {
+		return MembershipOrOwnershipTypeNone, trace.AccessDenied("User '%s' is an owner, but does not have the roles or traits required to be an owner of list '%s'.", identity.Username, accessList.GetName())
+	}
```
Can this logic now be replaced by using Hierarchy?

r0mant commented on lib/services/access_list.go:
```diff
@@ -218,47 +233,165 @@ func UnmarshalAccessListMember(data []byte, opts ...MarshalOption) (*accesslist.
+// recursiveIsAccessListOwnerCheck checks if supplied tlsca.Identity is an inherited owner of the accessList.
+// Ownership is inherited when the tlsca.Identity is a Member of an Access List, up to the maximum allowed depth,
+// which is an explicit Owner of the provided accessList.
+func (a AccessListMembershipChecker) recursiveIsAccessListOwnerCheck(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) error {
```
This is now duplication of the same logic in Hierarchy - can we get rid of this?

kiosion replied:
Same as my reply above... in pursuit of keeping this PR from getting any bigger I'd much rather do a pass refactoring/removing these in a later PR, since the places they're used don't have access to `GetAccessLists`/`ListAccessLists`, and these instead make use of `GetAccessList`/`ListAccessListMembers`, so it'll take some refactoring.

r0mant commented on lib/services/access_list.go:
```diff
@@ -218,47 +233,165 @@ func UnmarshalAccessListMember(data []byte, opts ...MarshalOption) (*accesslist.
+//
+// If membership is inherited (through the tlsca.Identity having Membership in a nested Access List),
+// the tlsca.Identity must also meet the membership requirements of the nested list.
+func (a AccessListMembershipChecker) IsAccessListMember(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) (MembershipOrOwnershipType, error) {
```
Same question as above - the Hierarchy already implements this logic, can we reuse?

r0mant commented on lib/services/local/access_list.go:
```diff
@@ -303,6 +320,86 @@ func (a *AccessListService) GetAccessListMember(ctx context.Context, accessList
+// GetAccessListOwners returns a list of all owners in an Access List, including those inherited from nested Access Lists.
+//
+// Returned Owners are not validated for ownership requirements – use `IsAccessListOwner` for validation.
+func (a *AccessListService) GetAccessListOwners(ctx context.Context, accessListName string) ([]*accesslist.Owner, error) {
```
Similar question - can we take advantage of Hierarchy to simplify this function now?

kiosion replied:
Yup, done!

r0mant commented on lib/accesslists/hierarchy.go:
```diff
@@ -0,0 +1,569 @@
+	// Check explicit owners
+	if _, ok := node.OwnerUsers[user.GetName()]; ok {
+		// Verify ownership requirements using provided AccessList
+		if !userMeetsRequirements(user, node.AccessList.Spec.OwnershipRequires) {
```
Where do we validate owner/member TTL? If an owner/member TTL expired, they are no longer considered valid owner/member. I don't see the TTL check anywhere here, unless I missed it.

Same question for members function below.

kiosion replied:
Thanks for the catch, forgot to re-add that logic! Owners don't expire, but inherited Owners should be valid Members in their respective lists, so I added that logic in.

r0mant reviewed (commented):
@kiosion This is much closer to what I had in mind w/r/t hierarchy, the only main comment is now that we have it we should actually use it to simplify things in parts where now there's basically logic duplication, I left a few comments to that effect.

r0mant commented on lib/auth/userloginstate/generator.go:
```diff
@@ -162,43 +161,55 @@ func (g *Generator) Generate(ctx context.Context, user types.User) (*userloginst
-		if err := services.IsAccessListOwner(identity, accessList); err == nil {
-			g.grantRolesAndTraits(identity, accessList.Spec.OwnerGrants, state)
+		// Grants are inherited if the user is a member of the access list, explicitly or via inheritance.
+		if membershipKind, err := accessListHierarchy.IsAccessListMember(ctx, user, accessList.GetName()); err == nil && membershipKind != accesslists.MembershipOrOwnershipTypeNone {
```
Same question as in the other PR - can this be encapsulated by Hierarchy?

```go
roles, traits, err := accessListHierarchy.GetUserGrants(ctx, user, accessList.GetName())
```

The idea for it is that it encapsulates common functionality you need to do with nested access lists, not just ownership/membership checks.

kiosion replied:
Yeah, could do to move this into the hierarchy interface instead!

r0mant reviewed (commented):
I think we're getting close, just a couple of comments left on this one.

smallinsky reviewed (commented):
I have left a few comments. Once they are addressed, this looks good to me. Great work!

kiosion commented on lib/accesslists/hierarchy.go:
```diff
@@ -0,0 +1,829 @@
+	return maxDepth
+}
+
+// IsDescendant determines if a maybeChild AccessList is a descendant of a parent AccessList.
+func (h *hierarchy) IsDescendant(parentListName, maybeChildListName string, kind RelationshipKind) (bool, error) {
```
probably not - I left it as part of the interface before moving `GetInheritedGrants` into the interface, it could probably be private now.

kiosion commented on lib/accesslists/hierarchy.go:
```diff
@@ -0,0 +1,829 @@
+		if member, ok := memberList.MemberUsers[user.GetName()]; ok {
+			// Check if membership is expired
+			if !member.Spec.Expires.IsZero() && !h.Clock.Now().Before(member.Spec.Expires) {
+				expired = true
```
thanks for the catch! missed transferring that logic over. should be fixed now.

strideynet commented on api/types/accesslist/convert/v1/accesslist.go:
```diff
@@ -209,11 +209,51 @@ func ToProto(accessList *accesslist.AccessList) *accesslistv1.AccessList {
+			ineligibleStatus = accesslistv1.IneligibleStatus(enumVal)
+		}
+	} else {
+		ineligibleStatus = accesslistv1.IneligibleStatus_INELIGIBLE_STATUS_UNSPECIFIED
```
Nit: This is probably unnecessary. The zero value of ineligibleStatus should already be IneligibleStatus_INELIGIBLE_STATUS_UNSPECIFIED.

strideynet commented on lib/services/local/access_list.go:
```diff
@@ -303,6 +321,30 @@ func (a *AccessListService) GetAccessListMember(ctx context.Context, accessList
+// GetAccessListOwners returns a list of all owners in an Access List, including those inherited from nested Access Lists.
+//
+// Returned Owners are not validated for ownership requirements – use `IsAccessListOwner` for validation.
+func (a *AccessListService) GetAccessListOwners(ctx context.Context, accessListName string) ([]*accesslist.Owner, error) {
```
Nit: There's a few potential sources of errors w/in this function. Introducing error wrapping will help us pin-point the source in future, e.g:

```
	accessLists, err := a.GetAccessLists(ctx)
	if err != nil {
		return nil, trace.Wrap(err, "getting access lists")
	}
	hierarchy, err := accesslists.NewHierarchy(ctx, accesslists.HierarchyConfig{
		AccessLists: accessLists,
		Members:     a,
		Locks:       nil,
		Clock:       a.clock,
	})
	if err != nil {
		return nil, trace.Wrap(err, "creating hierarchy")
	}
	owners, err := hierarchy.GetOwners(accessListName)
	if err != nil {
		return nil, trace.Wrap(err, "getting owners")
	}
```

r0mant reviewed (commented):
@kiosion A few more fairly minor comments, after that should be lgtm.

kiosion commented on lib/auth/userloginstate/generator.go:
```diff
@@ -208,13 +224,19 @@ func (g *Generator) grantRolesAndTraits(identity tlsca.Identity, grants accessli
+
+	// Return inherited roles and traits if the user is a member/owner through inheritance.
+	if membershipOrOwnershipType == accesslists.MembershipOrOwnershipTypeInherited {
+		return grants.Roles, grants.Traits
```
Yeah, this bit of logic could be simplified quite a bit. Not sure why I did it like this to begin with...

kiosion committed:
feat: Allow including Access Lists as nested members and owners

- Recursively check for accesslist membership

- Allow adding/removing/listing included access lists in acl commands

- Add a recursive test

- Use dynamic access lists structure from RFD

- Resolve proto changes

- Exclude 'list' members from Access List memberCount

- Calc Access List member count with members of type 'list' excluded,
  return seperately to front end

- Update examples/integrations

- Update crd docs

- Update tf docs

- Perform calculation of inherited roles/traits to AccessList service in
  order to utilize cache and minimize number of requests.

- Grant Okta integration RO for Access Lists

- Update AccessListMember-* events

- Include count for inherited grants

- Include MembershipKind of affected member(s)

- Emit inherited grants / members' MembershipKind for AccessListMember-*
  events

- Update notified owners for Access Requests

- Ensure dynamic owners are notified for Access Requests

- Ensure dynamic owners are notified via Slack integration

- Optionally pass an AbortSignal to `fetchAccessLists` in Web UI

- Replace usages of `services.IsAccessListOwner/IsAccessListMember` with
  equivelant funcs from `Hierarchy`

- Remove final references to AccessListMembershipChecker

- Don't allow ACL deletion when member/owner in other lists

- Guard Access List deletion behind membership/ownership checks for List

- Expose Hierarchy func to recursively get all members

- Tidy UserLoginStateGenerator logic involving ACL Membership/Ownership

r0mant commented on lib/accesslists/hierarchy.go:
```diff
@@ -0,0 +1,837 @@
+	Clock clockwork.Clock
+}
+
+type Hierarchy interface {
```
Nit: Feels like this being the central interface for interacting with nested access lists should have a good godoc on it.

kiosion replied:
added one, WDYT? the funcs have more detailed docs themselves, not sure what else to add to the interface doc

r0mant replied:
Looks good to me!

r0mant commented on lib/accesslists/hierarchy.go:
```diff
@@ -0,0 +1,837 @@
+}
+
+func checkAndSetDefaults(cfg *HierarchyConfig) error {
+	if cfg.AccessLists == nil {
```
Is empty hierarchy valid? Should probably be, just wondering.

kiosion replied:
Hm, yeah, it should be, e.g. if you're creating your first Access List. Not sure why I spaced on that.

kiosion replied:
Although, does Go treat an empty slice as `nil`? I assume if you simply have no existing lists, calling `cache.GetAccessLists(ctx)` wouldn't return `nil`, but I could be wrong...

r0mant reviewed (approved):
lgtm :shipit: Thank you for putting up with all my feedback @kiosion :)

fspmarshall reviewed (changes requested):
As currently implemented, it is possible to introduce a cycle by concurrently creating a recursive relationship.  I manually replicated this using https://github.com/gravitational/teleport/commit/23955957c885946edf24723845d2372ae529d434.

Once a cycle is introduced into the backend, the cluster is effectively bricked because it's impossible to login since the access lists can't be evaluated.

More generally, cross-list reference validation isn't being performed in a concurrency-safe manner.  Given the severity of broken/illegal linkages, I think we need to be really really confident in the concurrency-safety of validation.   IMO even `RunWhileLocked` likely isn't sufficient, and a solution based on `AtomicWrite` is probably what we want.

kiosion commented on lib/auth/userloginstate/generator_test.go:
```diff
@@ -375,9 +405,148 @@ func TestAccessLists(t *testing.T) {
+			expectedRoleCount:           1,
+			expectedTraitCount:          1,
+			expectedInheritedRoleCount:  1,
+			expectedInheritedTraitCount: 1,
```
true, can add a few of those cases.

kiosion commented on lib/services/local/access_list.go:
```diff
@@ -161,73 +186,195 @@ func (a *AccessListService) GetAccessListsToReview(ctx context.Context) ([]*acce
+			return trace.Wrap(err)
+		}
+
+		// Check if the access list is a member or owner of any other access lists.
+		if len(accessList.Spec.MemberOf) > 0 {
+			return trace.AccessDenied("Cannot delete '%s', as it is a member of one or more other Access Lists", accessList.Spec.Title)
+		}
+		if len(accessList.Spec.OwnerOf) > 0 {
+			return trace.AccessDenied("Cannot delete '%s', as it is an owner of one or more other Access Lists", accessList.Spec.Title)
+		}
```
good idea... I do still feel a bit unsure about those fields. I've tested every scenario I can think of and haven't seen them get out of whack, but I'm worried it *could* happen. is there a better way we could be keeping them updated, maybe listening for AccessListMemberCreated events or something instead?

kiosion replied:
nevermind... found a few cases where they can get out of sync 😓 

fixing those now, I also think not erroring if they're updated is best, and just ignoring the passed-in values, and going with what's already saved (if updating), otherwise explicitly setting to empty arrays (if creating new list).
